### PR TITLE
8320798: Console read line with zero out should zero out underlying buffer

### PR DESCRIPTION
**Pull request by naotoj**
It is best practice to zero out the underlying buffer after use.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320798](https://bugs.openjdk.org/browse/JDK-8320798): Console read line with zero out should zero out underlying buffer (**Bug** - P4)

### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to [90c4dece](https://git.openjdk.org/jdk/pull/16861/files/90c4dece613ae5d154da09dfbbeb8475c743a79b)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**) ⚠️ Review applies to [90c4dece](https://git.openjdk.org/jdk/pull/16861/files/90c4dece613ae5d154da09dfbbeb8475c743a79b)
 * [Joe Wang](https://openjdk.org/census#joehw) (@JoeWang-Java - **Reviewer**) ⚠️ Review applies to [90c4dece](https://git.openjdk.org/jdk/pull/16861/files/90c4dece613ae5d154da09dfbbeb8475c743a79b)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [90c4dece](https://git.openjdk.org/jdk/pull/16861/files/90c4dece613ae5d154da09dfbbeb8475c743a79b)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16861/head:pull/16861` \
`$ git checkout pull/16861`

Update a local copy of the PR: \
`$ git checkout pull/16861` \
`$ git pull https://git.openjdk.org/jdk.git pull/16861/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16861`

View PR using the GUI difftool: \
`$ git pr show -t 16861`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16861.diff">https://git.openjdk.org/jdk/pull/16861.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16861#issuecomment-1830522005)

### CODE REVIEW

File: src/java.base/share/classes/sun/nio/cs/StreamDecoder.java [hunk 1]
```diff
@@ -271,6 +273,26 @@ private boolean isOpen() {
...
+        } else {
+            synchronized (lock) {
+                lockedFillZeroToPosition();
+            }
+        }
+    }
+
+    private void lockedFillZeroToPosition() throws IOException {
+        ensureOpen();
+        Arrays.fill(bb.array(), 0, bb.arrayOffset() + bb.position(), (byte)0);
```
Comment (bplb) [inline src/java.base/share/classes/sun/nio/cs/StreamDecoder.java:294]: Should the `from` parameter be zero or `bb.arrayOffset()`?

Comment (naotoj) [reply src/java.base/share/classes/sun/nio/cs/StreamDecoder.java]: Right. Fixed.

**Review by bplb — APPROVED** [review]
Looks good now.

**Review by LanceAndersen — APPROVED** [review]

**Review by JoeWang-Java — APPROVED** [review]

File: src/java.base/share/classes/sun/nio/cs/StreamDecoder.java (hunk 1 above)
Comment (AlanBateman) [reply src/java.base/share/classes/sun/nio/cs/StreamDecoder.java]: The buffer is allocated in StreamDecoder so I assume the array offset is 0 anyway, so I think they will be the same.

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/share/classes/sun/nio/cs/StreamDecoder.java [hunk 2]
```diff
@@ -271,6 +273,26 @@ private boolean isOpen() {
...
+            }
+        } else {
+            synchronized (lock) {
+                lockedFillZeroToPosition();
+            }
+        }
+    }
+
+    private void lockedFillZeroToPosition() throws IOException {
+        ensureOpen();
```
Comment (jaikiran) [inline src/java.base/share/classes/sun/nio/cs/StreamDecoder.java:293]: Hello Naoto, is this open check needed? Given that we are just zeroing out already read content, would it thus be OK to clear that content to zero even after this `Reader` instance has been closed?

**Comment by AlanBateman** [comment]
> Given that the default `JdkConsole` provider implementation is the one that uses jline (in `jdk.internal.le` module), should we also do something to zero out this data in its implementation too. A brief glance suggests that it too uses an internal implementation class called `jdk.internal.org.jline.reader.impl.BufferImpl`.

This PR is the console implementation in java.base, there will be other changes for jline.

File: src/java.base/share/classes/sun/nio/cs/StreamDecoder.java (hunk 2 above)
Comment (naotoj) [reply src/java.base/share/classes/sun/nio/cs/StreamDecoder.java]: Thanks, Jai. Removed it.

**Review by jaikiran — APPROVED** [review]
Thank you for the update Naoto. The changes look good to me.

**Comment by naotoj** [comment]
Thanks all for the reviews!

/integrate

**Pull request closed without merging**